### ISSUE TITLE
Check failed: error == cudaSuccess (29 vs. 0)  driver shutting down

### ISSUE THREAD

**Issue by deepali-c** [NVIDIA/caffe#267]
I am using Caffe v 0.15.13 and I have trained an alexnet model using my own dataset. 

Next when I am trying to test the model using the command:
`caffe test -weights <path to my model> -model <path to my train_val.prototxt> -iterations 1 `

I observe the following error:
```
F1114 11:21:01.718121  4386 gpu_memory.cpp:136] Check failed: error == cudaSuccess (29 vs. 0)  driver shutting down
*** Check failure stack trace: ***
    @     0x3fffaab6ce0c  google::LogMessage::Fail()
    @     0x3fffaab6f284  google::LogMessage::SendToLog()
    @     0x3fffaab6c768  google::LogMessage::Flush()
    @     0x3fffaab711c4  google::LogMessageFatal::~LogMessageFatal()
    @     0x3fffaaeaae50  caffe::GPUMemory::Manager::deallocate()
    @     0x3fffaaead6ac  boost::detail::sp_counted_impl_p<>::dispose()
    @     0x3fffaaf13a88  caffe::GPUMemory::MultiWorkspace::~MultiWorkspace()
    @     0x3fffaa263d0c  __cxa_finalize
    @     0x3fffaad8ee80  (unknown)
    @     0x3fffab71c3c8  (unknown)
    @     0x3fffaa26377c  (unknown)
    @     0x3fffaa2637c4  exit
    @     0x3fffaa243124  (unknown)
    @     0x3fffaa243338  __libc_start_main
    @              (nil)  (unknown)
Aborted (core dumped)
```

**Comment by deepali-c** [comment]
Similar crash is seen when using **pre-trained bvlc_alexnet** model. 

Sample command:
`caffe test -weights ../models/bvlc_alexnet/bvlc_alexnet.caffemodel -model ../models/bvlc_alexnet/train_val.prototxt -iterations 1`

The following error is printed towards the end and core is generated:

```
F1116 12:13:04.339265  2616 gpu_memory.cpp:136] Check failed: error == cudaSuccess (29 vs. 0)  driver shutting down
*** Check failure stack trace: ***
    @     0x3fffa7bece0c  google::LogMessage::Fail()
    @     0x3fffa7bef284  google::LogMessage::SendToLog()
    @     0x3fffa7bec768  google::LogMessage::Flush()
    @     0x3fffa7bf11c4  google::LogMessageFatal::~LogMessageFatal()
    @     0x3fffa7f2ae50  caffe::GPUMemory::Manager::deallocate()
    @     0x3fffa7f2d6ac  boost::detail::sp_counted_impl_p<>::dispose()
    @     0x3fffa7f93a88  caffe::GPUMemory::MultiWorkspace::~MultiWorkspace()
    @     0x3fffa72e3d0c  __cxa_finalize
    @     0x3fffa7e0ee80  (unknown)
    @     0x3fffa879c3c8  (unknown)
    @     0x3fffa72e377c  (unknown)
    @     0x3fffa72e37c4  exit
    @     0x3fffa72c3124  (unknown)
    @     0x3fffa72c3338  __libc_start_main
    @              (nil)  (unknown)
Aborted (core dumped)
```

**Comment by lukeyeager** [comment]
I don't get an error. Here's the command I used:

```
./build/tools/caffe test -iterations 1 -gpu 0,1 \
    -model   ./models/bvlc_alexnet/train_val.prototxt \
    -weights ./models/bvlc_alexnet/bvlc_alexnet.caffemodel
```

Last few lines are:

```
I1116 09:06:57.657089 17279 caffe.cpp:277] Running for 1 iterations.
I1116 09:06:57.729657 17279 caffe.cpp:300] Batch 0, accuracy = 0.48
I1116 09:06:57.729686 17279 caffe.cpp:300] Batch 0, loss = 1.99947
I1116 09:06:57.729691 17279 caffe.cpp:305] Loss: 1.99947
I1116 09:06:57.729697 17279 caffe.cpp:317] accuracy = 0.48
I1116 09:06:57.729708 17279 caffe.cpp:317] loss = 1.99947 (* 1 = 1.99947 loss)
```

I've seen "driver shutting down" somewhere before. Seems like a problem with your driver or CUDA toolkit rather than with your Caffe build.

**Comment by drnikolaev** [comment]
https://github.com/NVIDIA/caffe/blob/caffe-0.15/src/caffe/util/gpu_memory.cpp#L136

Line 136 has no code. Please check the version you build. There was such bug but it's now fixed in 0.15

**Comment by deepali-c** [comment]
Just want to share an update here. 

I tried the command with the latest code on master branch and it does work. However, when I use code corresponding to `v0.15.13`, it crashes. 

[Line 136](https://github.com/NVIDIA/caffe/blob/v0.15.13/src/caffe/util/gpu_memory.cpp#L136) has the code -`CUDA_CHECK(cudaFree(ptr));` in `v0.15.13`. 

Just adding code corresponding to the commit  [Add support for GPU transforms to reduce CPU load](https://github.com/NVIDIA/caffe/commit/5f2b719d49cb3a225befcb532b47c86f2b5afd2f) into `v0.15.13` code helps in resolving the crash.  This change however does not affect `src/caffe/util/gpu_memory.cpp` but it still works.
